0030-simple-classical-representations.md

```diff
@@ -0,0 +1,260 @@
+2. indexing the bits resulting from one of these rvalue expresions;
+3. slicing a `uint[n]` to retrieve a subset of its bits;
+4. any method of assigning one of these expressions for reuse elsewhere in the circuit.
+
+Points 1 through 3 are rejected in this initial proposal for the sake of keeping the number of expression-tree nodes and type-system entries small.
+Point 4 is rejected because it would require the definition of additional `Operation` objects and potentially modifications to the `QuantumCircuit` class that we are not yet prepared to make.
+All of these points are eligible to be included in future work, just not the MVP.
+
+
+### Construction
+
+#### Stage 1
+
+This is a temporary representation, so it is a non-goal to produce the best possible UX for constructing these objects.
+In the initial representation phase, users will be required to build the expression-tree representations themselves.
+This will look _something_ like (names up for discussion):
+
+```python
+from qiskit.circuit import Clbit, ClassicalRegister
+from qiskit.circuit.classical import expr
+
+loose = [Clbit() for _ in [None]*3]
+cr1 = ClassicalRegister(2)
+cr2 = ClassicalRegister(2)
+
+# Equivalent to the current condition `(cr, 2)`:
+equal_reg = expr.Equal(expr.Value(cr1), expr.Value(2))
+
+# Equivalent to the current condition `(loose[0], False)`
+equal_bit = expr.Equal(expr.Value(loose[0]), expr.Value(False))
+
+# The above example, but using a Bool-valued unary expression:
+not_bit = expr.Not(expr.Value(loose[0]))
+
+# The comparison 0 < cr1 <= cr2
+bounded_reg = expr.And(
+    expr.Less(expr.Value(0), expr.Value(cr1),
+    expr.LessEqual(expr.Value(cr1), expr.Value(cr2),
+)
+```
+
+These expressions will then be given directly to `IfElseOp` or `WhileLoopOp` in their `condition` fields.
+These expressions will not be valid in the general `Instruction.condition` field; we do not want to expand any support for something we are already moving to remove.
+
+
+#### Stage 2
+
+Looking beyond the initial phase, it will be convenient to define a less verbose way for users to create these expressions.
+Since this document is intended to define a representation that can be removed and deprecated from Terra easily if it needs to evolve, we will _not_ consider overloading the Python magic-methods `__and__` etc on `Clbit` and `ClassicalRegister` in general Terra usage.
+However, we can do so in a limited scope, which will temporarily change the semantics.
+This will look like (continued from above code block, and the same examples in the same order):
+
+```python
+with expr.build():
+    equal_reg = cr1 == 2
+    equal_bit = loose[0] == False
+    not_bit = ~loose[0]
+    bounded_reg = (0 < cr1) & (cr1 <= cr2)
+```
+
+It is not possible to overload the behaviour of the Python Boolean operators (`not`, `and` and `or`) for classes, which is why the bitwise operations `~` and `&` are used instead (with the corresponding precedence frustrations).
```

Comment:
I could see this being a source of confusion. e.g. "Why can I `cr[0] == cr[1]` but not `cr[0] and cr[1]`?". Are there any cases where this could lead to errors like "I wanted a bitwise AND but got a logical AND"?

Reply:
The way I wrote this, you'll always have the option to use the regular constructors, or potentially Kevin's suggested chained methods, even in the middle of the builder context.  But yeah, it's not ideal, but imo magic methods give the best expression legibility in the majority of cases, so until we decide to commit to a _complete_ function-based builder interface where we can do AST inspection, this is my preferred choice.

Also: my intent was for `&` to still be bitwise AND, it's just also defined on bools.  You'd have to be explicit to get logical operations (with the corresponding short-circuiting)